File: src/controleurs/scripts/rail_mover.py
#!/usr/bin/env python3
"""
rail_mover.py
=============
Déplace le robot et le curseur sur le rail avec interpolation lisse.

Architecture
------------
  - Boucle de contrôle à 30 Hz (STEP_DT = 0.033 s) dans un thread dédié.
  - Les appels Gazebo (set_pose) sont délégués à un worker thread via une
    queue latest-value : si le worker est occupé, la nouvelle position
    écrase l'ancienne en attente — jamais de backlog.
  - Robot et curseur sont toujours déplacés ensemble (même Y, même appel).

Topics
------
  SUB  /target_rail_pos          std_msgs/Float64   position cible (m)
  PUB  /rail_mover/done          std_msgs/Bool      True = arrivée confirmée
  PUB  /current_rail_position    std_msgs/Float64   position interpolée courante

Paramètres
----------
  initial_rail_position  float  défaut 0.0    position de départ (m)
  rail_speed             float  défaut 0.35   vitesse de déplacement (m/s)
"""

import time
import queue
import threading
import subprocess

import rclpy
from rclpy.node import Node
from std_msgs.msg import Bool, Float64


# ── Constantes géométriques ────────────────────────────────────────────────
Y_BASE = -0.85   # y robot quand rail_position = 0.0
Z_ARM  =  0.0    # z origine probo11
Z_CURS =  1.03   # z curseur

# ── Boucle de contrôle ─────────────────────────────────────────────────────
STEP_DT  = 0.033  # s  — période de mise à jour de la position (≈30 Hz)
GOAL_TOL = 1e-3   # m  — seuil d'arrivée


# ══════════════════════════════════════════════════════════════════════════
# Worker Gazebo : queue latest-value
# ══════════════════════════════════════════════════════════════════════════

class GazeboPoseWorker:
    """
    Thread unique qui appelle 'ign service set_pose' en continu.
    Si une nouvelle position arrive pendant qu'un appel est en cours,
    elle remplace la précédente en attente (latest-value queue, taille 1).
    Robot et curseur sont toujours déplacés simultanément (deux Popen
    lancés en parallèle, pas de wait() bloquant la boucle de contrôle).
    """

    def __init__(self, logger=None):
        self._logger  = logger
        self._q       = queue.Queue(maxsize=1)
        self._thread  = threading.Thread(target=self._loop, daemon=True)
        self._thread.start()

    def send(self, y: float):
        """Envoie une position. Écrase l'ancienne si le worker est encore occupé."""
        try:
            self._q.get_nowait()   # vide la queue si déjà une valeur en attente
        except queue.Empty:
            pass
        self._q.put(y)

    def send_blocking(self, y: float) -> bool:
        """
        Appel bloquant final : attend la confirmation Gazebo avant de retourner.
        Utilisé pour la dernière position (arrivée) afin de garantir que
        Gazebo a bien reçu la position finale avant de signaler 'done'.
        """
        return self._call_both(y)

    # ── Interne ─────────────────────────────────────────────────────────
    def _loop(self):
        while True:
            y = self._q.get()          # attend une position
            self._call_both(y)         # appel bloquant (OK, on est dans un thread)

    def _call_both(self, y: float) -> bool:
        """Lance robot + curseur en parallèle, attend les deux."""
        results = [False, False]

        def robot():
            results[0] = self._set_pose('probo11',      y, Z_ARM)

        def cursor():
            results[1] = self._set_pose('rail_curseur', y, Z_CURS)

        t1 = threading.Thread(target=robot,  daemon=True)
        t2 = threading.Thread(target=cursor, daemon=True)
        t1.start(); t2.start()
        t1.join();  t2.join()
        return results[0] and results[1]

    def _set_pose(self, name: str, y: float, z: float) -> bool:
        req = (f'name: "{name}" '
               f'position: {{x: 0.0000, y: {y:.4f}, z: {z:.4f}}}')
        r = subprocess.run(
            ['ign', 'service',
             '-s', '/world/empty/set_pose',
             '--reqtype', 'ignition.msgs.Pose',
             '--reptype', 'ignition.msgs.Boolean',
             '--timeout', '1500',
             '--req', req],
            capture_output=True, text=True
        )
        return r.returncode == 0


# ══════════════════════════════════════════════════════════════════════════
# Nœud ROS 2
# ══════════════════════════════════════════════════════════════════════════

class RailMover(Node):

    def __init__(self):
        super().__init__('rail_mover')

        self.declare_parameter('initial_rail_position', 0.0)
        self.declare_parameter('rail_speed', 0.35)

        self.rail_pos_   = self.get_parameter(
            'initial_rail_position').get_parameter_value().double_value
        self.rail_speed_ = self.get_parameter(
            'rail_speed').get_parameter_value().double_value

        self._target     = self.rail_pos_
        self._moving     = False
        self._lock       = threading.Lock()
        self._worker     = GazeboPoseWorker(self.get_logger())

        self.pub_done_ = self.create_publisher(Bool,    '/rail_mover/done',       10)
        self.pub_pos_  = self.create_publisher(Float64, '/current_rail_position', 10)
        self.create_subscription(Float64, '/target_rail_pos', self._on_target, 5)

        self._publish_current()
        self.get_logger().info(
            f'rail_mover prêt — pos={self.rail_pos_:.3f} m  '
            f'vitesse={self.rail_speed_:.2f} m/s')

    # ── Callback cible ──────────────────────────────────────────────────

    def _on_target(self, msg: Float64):
        new = max(0.0, min(1.7, float(msg.data)))

        if abs(new - self.rail_pos_) < GOAL_TOL:
            self.get_logger().info('Rail déjà en position.')
            self._publish_done(True)
            return

        with self._lock:
            self._target = new
            if not self._moving:
                self._moving = True
                threading.Thread(target=self._move_loop, daemon=True).start()

    # ── Boucle d'interpolation ──────────────────────────────────────────

    def _move_loop(self):
        with self._lock:
            target = self._target

        self.get_logger().info(
            f'Rail : {self.rail_pos_:.3f} → {target:.3f} m  '
            f'(v={self.rail_speed_:.2f} m/s)')

        while True:
            with self._lock:
                target = self._target          # cible éventuellement mise à jour

            remaining = target - self.rail_pos_
            if abs(remaining) <= GOAL_TOL:
                break

            step = self.rail_speed_ * STEP_DT
            if abs(remaining) <= step:
                self.rail_pos_ = target
            else:
                self.rail_pos_ += step * (1.0 if remaining > 0.0 else -1.0)

            # Envoi non-bloquant → la boucle de contrôle n'attend pas Gazebo
            self._worker.send(Y_BASE + self.rail_pos_)
            self._publish_current()

            time.sleep(STEP_DT)

        # ── Arrivée : appel bloquant pour confirmer la position finale ──
        self.rail_pos_ = target
        ok = self._worker.send_blocking(Y_BASE + self.rail_pos_)
        self._publish_current()

        with self._lock:
            self._moving = False

        if ok:
            self.get_logger().info(f'Rail en position : {self.rail_pos_:.3f} m')
        else:
            self.get_logger().warn('set_pose final incertain — Gazebo occupé ?')

        # done=True seulement après confirmation Gazebo de la position finale
        self._publish_done(True)

    # ── Helpers ────────────────────────────────────────────────────────

    def _publish_done(self, v: bool):
        msg = Bool(); msg.data = v; self.pub_done_.publish(msg)

    def _publish_current(self):
        msg = Float64(); msg.data = self.rail_pos_; self.pub_pos_.publish(msg)


def main():
    rclpy.init()
    rclpy.spin(RailMover())
    rclpy.shutdown()


if __name__ == '__main__':
    main()

File: src/controleurs/src/rail_mover.cpp
/**
 * rail_mover.cpp
 * ==============
 * Déplace le robot (probo11) et le curseur (rail_curseur) sur le rail
 * via ignition::transport directement — sans subprocess, sans latence.
 *
 * À 100 Hz, entre deux appels set_pose : chute gravité = 0.5×9.81×(0.01)² ≈ 0.5 mm.
 *
 * Topics ROS 2
 * ------------
 *   SUB  /target_rail_pos          std_msgs/Float64   position cible (m)
 *   PUB  /rail_mover/done          std_msgs/Bool      True = arrivée
 *   PUB  /current_rail_position    std_msgs/Float64   position interpolée
 *
 * Paramètres
 * ----------
 *   initial_rail_position  double  défaut 0.0    position de départ (m)
 *   rail_speed             double  défaut 0.35   vitesse (m/s)
 *   world_name             string  défaut "empty" nom du world Gazebo
 */

#include <rclcpp/rclcpp.hpp>
#include <std_msgs/msg/bool.hpp>
#include <std_msgs/msg/float64.hpp>

#include <ignition/transport/Node.hh>
#include <ignition/msgs/pose.pb.h>
#include <ignition/msgs/boolean.pb.h>

#include <chrono>
#include <cmath>
#include <string>

using namespace std::chrono_literals;

// ── Géométrie ────────────────────────────────────────────────────────────────
static constexpr double Y_BASE  = -0.85;   // y robot quand rail_position = 0
static constexpr double Z_ARM   =  0.0;    // z origine probo11
static constexpr double Z_CURS  =  1.03;   // z curseur
static constexpr double GOAL_TOL = 1e-3;   // seuil d'arrivée (m)

// ════════════════════════════════════════════════════════════════════════════
class RailMover : public rclcpp::Node
{
public:
    RailMover() : Node("rail_mover"), moving_(false)
    {
        // ── Paramètres ───────────────────────────────────────────────────
        this->declare_parameter("initial_rail_position", 0.0);
        this->declare_parameter("rail_speed", 0.35);
        this->declare_parameter("world_name", std::string("empty"));

        rail_pos_   = this->get_parameter("initial_rail_position")
                          .get_parameter_value().get<double>();
        rail_speed_ = this->get_parameter("rail_speed")
                          .get_parameter_value().get<double>();
        world_name_ = this->get_parameter("world_name")
                          .get_parameter_value().get<std::string>();

        target_pos_ = rail_pos_;
        service_name_ = "/world/" + world_name_ + "/set_pose";

        // ── Publishers ───────────────────────────────────────────────────
        pub_done_ = create_publisher<std_msgs::msg::Bool>(
            "/rail_mover/done", 10);
        pub_pos_  = create_publisher<std_msgs::msg::Float64>(
            "/current_rail_position", 10);

        // ── Subscriber ───────────────────────────────────────────────────
        sub_target_ = create_subscription<std_msgs::msg::Float64>(
            "/target_rail_pos", 5,
            [this](const std_msgs::msg::Float64::SharedPtr msg) {
                onTarget(msg->data);
            });

        // ── Timer de contrôle à 100 Hz ───────────────────────────────────
        timer_ = create_wall_timer(10ms,
            [this]() { controlLoop(); });

        publishCurrent();
        RCLCPP_INFO(get_logger(),
            "rail_mover prêt — pos=%.3f m  vitesse=%.2f m/s  service=%s",
            rail_pos_, rail_speed_, service_name_.c_str());
    }

private:
    // ── Callback cible ───────────────────────────────────────────────────────
    void onTarget(double new_pos)
    {
        new_pos = std::max(0.0, std::min(1.7, new_pos));

        if (std::abs(new_pos - rail_pos_) < GOAL_TOL) {
            RCLCPP_INFO(get_logger(), "Rail déjà en position.");
            publishDone(true);
            return;
        }

        RCLCPP_INFO(get_logger(),
            "Rail : %.3f → %.3f m  (v=%.2f m/s)",
            rail_pos_, new_pos, rail_speed_);

        target_pos_ = new_pos;
        moving_     = true;
    }

    // ── Boucle de contrôle (100 Hz) ──────────────────────────────────────────
    void controlLoop()
    {
        // Même à l'arrêt : on maintient la position pour contrecarrer la gravité.
        if (!moving_) {
            setPoseBoth(Y_BASE + rail_pos_);
            return;
        }

        double remaining = target_pos_ - rail_pos_;

        if (std::abs(remaining) <= GOAL_TOL) {
            // Arrivée — appel final bloquant pour confirmer
            rail_pos_ = target_pos_;
            setPoseBoth(Y_BASE + rail_pos_);
            publishCurrent();
            moving_ = false;
            RCLCPP_INFO(get_logger(), "Rail en position : %.3f m", rail_pos_);
            publishDone(true);
            return;
        }

        double step = rail_speed_ * 0.01;   // v × dt (dt = 10 ms)
        if (std::abs(remaining) <= step)
            rail_pos_ = target_pos_;
        else
            rail_pos_ += step * (remaining > 0.0 ? 1.0 : -1.0);

        setPoseBoth(Y_BASE + rail_pos_);
        publishCurrent();
    }

    // ── set_pose via ignition transport (non bloquant côté timer) ────────────
    void setPoseBoth(double y)
    {
        // Les deux appels sont faits en séquence — chacun <5 ms via ign transport.
        setPose("probo11",      y, Z_ARM);
        setPose("rail_curseur", y, Z_CURS);
    }

    bool setPose(const std::string & model_name, double y, double z)
    {
        ignition::msgs::Pose req;
        req.set_name(model_name);
        req.mutable_position()->set_x(0.0);
        req.mutable_position()->set_y(y);
        req.mutable_position()->set_z(z);
        // Orientation identité (quaternion w=1)
        req.mutable_orientation()->set_w(1.0);
        req.mutable_orientation()->set_x(0.0);
        req.mutable_orientation()->set_y(0.0);
        req.mutable_orientation()->set_z(0.0);

        ignition::msgs::Boolean rep;
        bool result = false;
        // Timeout 200 ms — bien suffisant pour un appel local
        bool ok = ign_node_.Request(service_name_, req, 200, rep, result);
        if (!ok || !result) {
            RCLCPP_WARN_THROTTLE(get_logger(), *get_clock(), 2000,
                "set_pose(%s) échec — Gazebo prêt ?", model_name.c_str());
        }
        return ok && result;
    }

    // ── Helpers ─────────────────────────────────────────────────────────────
    void publishDone(bool v)
    {
        std_msgs::msg::Bool msg;
        msg.data = v;
        pub_done_->publish(msg);
    }

    void publishCurrent()
    {
        std_msgs::msg::Float64 msg;
        msg.data = rail_pos_;
        pub_pos_->publish(msg);
    }

    // ── Membres ──────────────────────────────────────────────────────────────
    ignition::transport::Node ign_node_;

    rclcpp::Publisher<std_msgs::msg::Bool>::SharedPtr    pub_done_;
    rclcpp::Publisher<std_msgs::msg::Float64>::SharedPtr pub_pos_;
    rclcpp::Subscription<std_msgs::msg::Float64>::SharedPtr sub_target_;
    rclcpp::TimerBase::SharedPtr timer_;

    double      rail_pos_;
    double      target_pos_;
    double      rail_speed_;
    bool        moving_;
    std::string world_name_;
    std::string service_name_;
};

// ════════════════════════════════════════════════════════════════════════════
int main(int argc, char ** argv)
{
    rclcpp::init(argc, argv);
    rclcpp::spin(std::make_shared<RailMover>());
    rclcpp::shutdown();
    return 0;
}
